### PR TITLE
Bootstrap `update nodegroup` command

### LINKED ISSUE #3827: Bootstrap new `update nodegroup` command
## New Feature: `eksctl update nodegroup`

The new `update nodegroup` command will be used for updating a nodegroup's configuration. 

`update nodegroup` is not to be confused with `upgrade nodegroup`, which is used for upgrading a nodegroup and its resources, for example to use a different AMI with a newer k8s version.

Instead, `update nodegroup` can be used to update configuration of nodegroups. This is important especially if the config needs to be changed prior to running an upgrade. 

Note that this is only available for managed nodegroups.

This PR relates to #3613 - it will be used to updating the new `UpdateConfig` field that will be used for handling parallel upgrades.

### PR DESCRIPTION
**Pull request by nikimanoledaki**
### Description

Closes #3827 

This PR bootstraps the `update nodegroup` command. 

Validates that:
- Nodegroup must be a managed nodegroup
- When used with a config file, a warning will log any unsupported fields that were present in the config file but could not be updated

```bash
./eksctl update nodegroup --config-file examples/15-managed-nodes.yaml
2021-06-16 15:49:20 [!]  unchanged fields: the following fields could not be updated because they are not supported by `eksctl update nodegroup`: InstanceType, AvailabilityZones, ScalingConfig, VolumeSize, SSH, Labels, Tags, IAM, SecurityGroups, Taints, UpdateConfig
2021-06-16 15:49:22 [ℹ]  nodegroup successfully updated
```

We don't want users to confused `upgrade` and `update` so every effort has been made to explain this in the `ctl` and user docs. We also don't want users to try to update fields in the config that cannot be updated through `update nodegroup` so a lot of validation was added to make sure only very specific fields in the config can be used with this command.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Made sure the title of the PR is a good description that can go into the release notes
- [x] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

### CODE REVIEW

File: userdocs/src/usage/eks-managed-nodes.md
Comment (aclevername) [inline userdocs/src/usage/eks-managed-nodes.md:188]: shall we add a list of the fields? or I guess in this PR there isn't any yet?

**Comment by aclevername** [comment]
@nikimanoledaki I think we don't want to merge this unless we have the PR that builds on top of this ready to get, otherwise we are shipping a new command that has no functionality :thinking:

File: userdocs/src/usage/eks-managed-nodes.md
Comment (nikimanoledaki) [reply userdocs/src/usage/eks-managed-nodes.md:188]: I had added a link to the example config but linter didn't like it because the link will appear broken until this PR is merged! https://github.com/weaveworks/eksctl/pull/3828/commits/4d888bc40944257b1a8e9bbbdb7a02779f262292

**Comment by nikimanoledaki** [comment]
@aclevername no problem! today I'll be making a branch with this one as its base to add `updateConfig` so that we can ~yeet~ merge both at the same time!

File: userdocs/src/usage/eks-managed-nodes.md
Comment (nikimanoledaki) [reply userdocs/src/usage/eks-managed-nodes.md:189]: I've made a note to make a one-liner PR after this one is merged to add the link to the example again, so that the linter doesn't complain :)

File: examples/29-update-nodegroup.yaml
Comment (Callisto13) [inline examples/29-update-nodegroup.yaml:10]: what's this example config demonstrating?

File: pkg/ctl/cmdutils/configfile.go
Comment (Callisto13) [inline pkg/ctl/cmdutils/configfile.go:779]: so just to clarify on the UX...

say I had a config file like
```yaml
managedNodeGroups:
- name: one
  someOtherField: foobar
- name: two
  updateConfig:
    shrugEmoji: true
- name: three
```
in order to run this command i would have to edit my config file to remove those nodegroups?

File: pkg/ctl/update/nodegroup_test.go
Comment (Callisto13) [inline pkg/ctl/update/nodegroup_test.go:56]: I may have missed something so this question may be irrelevant, but are we validating other fields for any other reason apart from clarity? Presumably we are only reading the `updateConfig` field for this command?

File: userdocs/src/usage/eks-managed-nodes.md
Comment (Callisto13) [reply userdocs/src/usage/eks-managed-nodes.md:189]: Can we add more to make it very clear that this will only ever update upgrade configuration?

I mean, who knows what we may use it for later, but let's not lead anyone on.

File: pkg/ctl/cmdutils/configfile.go
Comment (nikimanoledaki) [reply pkg/ctl/cmdutils/configfile.go:779]: right now yes but I'm not dead set on this UX, I meant to ask about this and I'm happy to change it. do we want `update nodegroup` to be able to update multiple nodegroups at once?

File: pkg/ctl/update/nodegroup_test.go
Comment (nikimanoledaki) [reply pkg/ctl/update/nodegroup_test.go:56]: it's just for clarity :)

File: pkg/ctl/cmdutils/configfile.go
Comment (Callisto13) [reply pkg/ctl/cmdutils/configfile.go:779]: I think so yes. People are likely to have their cluster config stored somewhere so it feels weird that we would require them to remove bits to run this command. With the config above I would have to comment-run-comment 3 times which does not seem fun.

File: pkg/ctl/update/nodegroup_test.go
Comment (Callisto13) [reply pkg/ctl/update/nodegroup_test.go:56]: cool cool

File: examples/29-update-nodegroup.yaml
Comment (nikimanoledaki) [reply examples/29-update-nodegroup.yaml:10]: nothing right now, this PR is just for bootstrapping `eksctl update nodegroup`! I'm creating another PR to use `update nodegroup` to update `updateConfig!

Comment (nikimanoledaki) [reply examples/29-update-nodegroup.yaml:10]: see: https://github.com/weaveworks/eksctl/pull/3855/commits/2f353953cddf57eb99fdc641a4941aacc2e7279b

File: pkg/ctl/cmdutils/configfile.go
Comment (nikimanoledaki) [reply pkg/ctl/cmdutils/configfile.go:779]: good shout! opened a new issue to add this afterwards: https://github.com/weaveworks/eksctl/issues/3857

File: userdocs/src/usage/eks-managed-nodes.md
Comment (nikimanoledaki) [reply userdocs/src/usage/eks-managed-nodes.md]: definitely, more info will be added in the PR that joins `update nodegroup` and `updateConfig`

**Comment by nikimanoledaki** [comment]
@Callisto13 I removed the validation that returned an error if there were `> 1` ng in the cfg file, so now the config file that's used for creating the cluster can be used! :) Will do the `multiple nodegroup` bit in a separate PR to keep things tidy. 

Also, it doesn't return errors anymore for unsupported fields and instead it just logs the fields that could not be updated (see description) ☺️ I like it! What do you think?

File: integration/tests/managed/managed_nodegroup_test.go
Comment (aclevername) [inline integration/tests/managed/managed_nodegroup_test.go:474]: whats this testing?

File: pkg/actions/nodegroup/update.go
Comment (aclevername) [inline pkg/actions/nodegroup/update.go:21]: slightly confusing because it could be either:

1. it doesn't exist
2. it exists as a `unmanaged` nodegroup

I think it might be worth distinguishing the two error cases, since this command might add a lot of confusion for users who try to update `unmanaged` nodegroups

File: userdocs/src/usage/eks-managed-nodes.md
Comment (aclevername) [inline userdocs/src/usage/eks-managed-nodes.md:200]: its none in this PR right? and then the first field is introduced in the follow PR? Might be worth moving the docs to the follow up PR

**Comment by Callisto13** [comment]
> I removed the validation that returned an error if there were > 1 ng in the cfg file, so now the config file that's used for creating the cluster can be used! 

So does this mean it will silently just act on the first in the array? Do we log that this is happening?

File: integration/tests/managed/managed_nodegroup_test.go
Comment (Callisto13) [reply integration/tests/managed/managed_nodegroup_test.go:474]: I believe it is testing that other config is ignored.

I know we don't normally do this and each command only looks at what is cares about, but since this is `update` and we often get asked things like "Can i update xyz on my ng without recreating?" it may make sense to be very clear and log the things in the config which are not going to be changed.

File: pkg/actions/nodegroup/update.go
Comment (Callisto13) [inline pkg/actions/nodegroup/update.go:26]: Where is the update call happening in this func? Am I being blind? Is github diff messing with me? 😂

Comment (Callisto13) [reply pkg/actions/nodegroup/update.go:26]: o wait is it happening in https://github.com/weaveworks/eksctl/pull/3855?

File: pkg/ctl/cmdutils/configfile.go
Comment (Callisto13) [inline pkg/ctl/cmdutils/configfile.go:809]: can we rename this `i`? bit of shadowing.

File: pkg/ctl/cmdutils/configfile.go
Comment (Callisto13) [inline pkg/ctl/cmdutils/configfile.go:793]: will this log even if that `unsupportedFields` array is empty?

File: integration/tests/managed/managed_nodegroup_test.go
Comment (nikimanoledaki) [reply integration/tests/managed/managed_nodegroup_test.go:474]: yes^ 👏 

I'd say informing the user of which fields were _not_ affected is so important for clarity of use that it's almost a feature and should be tested as such... it's part of making this command extra crystal clear to compensate for having a potentially confusing UX (`update` vs `upgrade`, `updateConfig` for upgrades)

Comment (nikimanoledaki) [reply integration/tests/managed/managed_nodegroup_test.go:474]: I'd (probably) be confused 😛

File: pkg/actions/nodegroup/update.go
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:21]: that's true, thank you, I'll find another way :)

File: pkg/actions/nodegroup/update.go
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:26]: yeah there's nothing happening here 🙈 all the fun happens over at #3855

File: pkg/ctl/cmdutils/configfile.go
Comment (nikimanoledaki) [reply pkg/ctl/cmdutils/configfile.go:793]: good shout, changing it to check `len` of `unsupportedFields` first!

File: pkg/actions/nodegroup/update.go
Comment (nikimanoledaki) [reply pkg/actions/nodegroup/update.go:21]: would it be less confusing and capture both error cases if the error message was simply reworded to `could not find managed nodegroup with name %q`?

File: userdocs/src/usage/eks-managed-nodes.md
Comment (nikimanoledaki) [reply userdocs/src/usage/eks-managed-nodes.md:200]: do you mean that we should move all user docs about `update` to the follow-up PR? the other PR is pretty much ready to go right after this one and they're meant to be released together so I don't think it makes much difference tbh, what do you think?

File: pkg/actions/nodegroup/update.go
Comment (aclevername) [reply pkg/actions/nodegroup/update.go:21]: > would it be less confusing and capture both error cases if the error message was reworded to could not find managed nodegroup with name %q?

sounds good!

File: userdocs/src/usage/eks-managed-nodes.md
Comment (aclevername) [reply userdocs/src/usage/eks-managed-nodes.md:200]: I think in that case lets leave it, but for the future If I was to go to this doc for a git blame, I wouldn't find the PR that actually introduced this functionality which would be slightly confusing, although we link to that PR in this PR so its not that bad :smile:

File: pkg/ctl/cmdutils/configfile.go
Comment (aclevername) [inline pkg/ctl/cmdutils/configfile.go:784]: do we not need to validate all managed nodegroups, since they are all getting updated?

**Review by aclevername — APPROVED** [review]
left another comment, but otherwise LGTM

File: pkg/ctl/cmdutils/configfile.go
Comment (nikimanoledaki) [reply pkg/ctl/cmdutils/configfile.go:784]: for now it's just 1 nodegroup but I'll open another PR right afterwards to iterate over all nodegroups! https://github.com/weaveworks/eksctl/pull/3828#discussion_r651658891

File: userdocs/src/usage/eks-managed-nodes.md
Comment (nikimanoledaki) [reply userdocs/src/usage/eks-managed-nodes.md:200]: ok I'll point to the other PR in this PR's merge commit!

**Comment by nikimanoledaki** [comment]
> So does this mean it will silently just act on the first in the array? Do we log that this is happening?

there's more logging in the other PR since not much happens in this one! sorry, it's a bit confusing!

**Pull request merged**
